### DISCUSSION TITLE
Support multiple consecutive underscores in an entity id

### DISCUSSION THREAD

**Discussion by lougreenwood** [home-assistant/architecture#900] [Entity Models] [answered]
I have devices which exists on the threshold of 2 rooms, I name these devices using the names of both rooms, to allow me to easily identify which specific device I'm working with I use the following format:

`Main Bedroom - Main Bedroom Ensuite Entrance Sensor`
`Main Bedroom Ensuite - Main Bedroom Entrance Sensor`
`Kitchen - Living Room Entrance Sensor`

If I didn't have the dash then the name is much harder to understand at a glance, for example `Main Bedroom Main Bedroom Ensuite Entrance Sensor` (where do I delineate which rooms this belongs to - I must read the whole name and infer it, which is cognitively expensive).

For these devices the entity id is generated to something like `binary_sensor.main_bedroom_main_bedroom_ensuite_entrance_motion`. Because of the limited valid chars for an entity id, I can't use a similar naming pattern.

I tried to use multiple consecutive underscores to signify the separator to make the entity id easier to read, but it seems that this is invalid (🤔), for example: `binary_sensor.main_bedroom__main_bedroom_ensuite_entrance_sensor`.

**So my question is:***

Is this choice to limit consecutive underscores by design?

I can't think of a reason why this might be an issue... well, actually I can - maybe it's somehow related to escaping the string, or maybe it's not valid in some intermediary format that HA needs to support, or maybe it's related to this: https://www.scaler.com/topics/double-underscore-python/# (in which case, validating that only the entity id doesn't start with a double underscore would be more appropriate?)

Is there an appetite to relax the validation to allow multiple consecutive underscores to provide more freedom to end users in naming patterns?

**Comment by frenck** [comment]
There is some history to be found around this:

- [Release notes Home Assistant 0.86](https://www.home-assistant.io/blog/2019/01/23/release-86/#noteworthy-breaking-changes)
- [_**Script id with '__' breaks configuration**_](https://github.com/home-assistant/core/issues/20100)
- [_**double underscores can crash HomeAssistant's startup**_](https://github.com/home-assistant/core/issues/20243)
- [_**Align valid_entity_id with new slugify**_](https://github.com/home-assistant/core/pull/20231)
- [_**Config Validator: schema_with_slug_keys**_](https://github.com/home-assistant/core/pull/20298)
- [_**Speed up validate_entity_id**_](https://github.com/home-assistant/core/pull/32137)

The main gist is that this is used for more than entity IDs.

I'm not entirely against the proposed change, but it would require more research from you to ensure the above-listed issues (and possibly new ones) are handled.

That said, of course, an entity ID cannot/may not start with a double underscore, but the relaxing could happen for mid-string.

What have been your findings thus far?

../Frenck

**Comment by lougreenwood** [comment]
Thanks @frenck - I only stumbled across this issue this morning. I'll do try to do some research and come back to this thread with some results.

I assume I would need to remove the validation in my local instance, start naming things with double underscore and then test against the issues described above to get some traction on this? Is that what you're looking for?

**Reply by frenck** [reply]
> Is that what you're looking for?

I guess what I'm looking for is research on the possible problems and risks that can occur by allowing for this.

The entity ID consists of those parts. A domain and an object ID, which separated by a dot. Both the domain and object IDs have wider use in Home Assistant (not just for an entity ID, but also for things like service names, references in storage collections, and for some - like scripts - even unique ID.

So the impact of changing this is not limited to entity IDs.

../Frenck

**Reply by C-Duv** [reply]
Let's hope this will get some traction because the `_` as the sole "universal" word separator is a bit painful when naming things, specially when we want to have a nice and organized naming scheme.

If allowing 2 `_` is difficult, maybe allowing another, dedicated character (`-`, `/`, `@`, `#`, `·`, …) would be less difficult?

**Comment by frenck** [comment, marked as answer]
We discussed this proposal in our architectural core meeting and decided not to proceed with it.

While it might be possible, we don't think we want to take the risks and effort involved for the arguable small gain this brings. Therefore, want to keep the existing boundaries of entity ID naming.

../Frenck
